### PR TITLE
Checking case completeness

### LINKED ISSUE #317: Case Expressions Have No Completeness Analysis
The following compiles without raising any errors or warnings:
```
case x of
| just(a) -> a
end
```
There should probably be some form of completeness analysis to ensure that all possible constructors are matched. This would simply require checking that all constructors from the language in which the nonterminal type is introduced are matched.

### PR DESCRIPTION
**Pull request by RandomActsOfGrammar**
This closes #317 by adding completeness analysis for case expressions.  If it finds a case expression's patterns are not complete, it generates an example pattern which is not covered by any of the current patterns.  For a set of patterns to be complete, it either needs to have a variable or wildcard case, or have all non-forwarding productions included in the patterns.

To support testing the completeness analysis, I added a testing production to check that code does not produce a warning containing certain text.

There are several incomplete matches still present in Silver.  I didn't do anything with them because I don't know why they are acceptable, so someone who does can fill them in with default cases of the form
```
| _ -> error("<why this case isn't needed>")
```
The same thing needs to be done for downstream projects, such as AbleC.  I have no experience with these projects, and therefore I can't give good reasons for why the incompleteness is acceptable.

# Documentation
Since we've been discussing adding a required documentation section for pull requests, I will do so here.

I have a number of comments explaining the functions I wrote for checking completeness; I believe these comments are sufficient for someone else to modify the code if necessary.  These do not require doc comments because they are internal to case analysis.

I have a doc comment for the testing production I added, since this is something others might wish to use.

I am working on adding a description of completeness analysis to the website.  This should be coming shortly.

### CODE REVIEW

File: grammars/silver/compiler/extension/implicit_monads/ProductionBody.sv
Comment (krame505) [inline grammars/silver/compiler/extension/implicit_monads/ProductionBody.sv:42]: Could use `fail.fromRight` here, instead.

File: grammars/silver/compiler/extension/patternmatching/Case.sv
Comment (krame505) [inline grammars/silver/compiler/extension/patternmatching/Case.sv:257]: Returning `Pair<Expr [Message]>` makes me suspicious that this would be better off as a pair of attributes on something (`AbstractMatchRules` if this was an actual nonterminal rather than a list?  IDK, but generally lists as abstract syntax trees are frowned upon.)  

But probably okay for now.

File: grammars/silver/compiler/extension/patternmatching/Case.sv
Comment (krame505) [inline grammars/silver/compiler/extension/patternmatching/Case.sv:426]: Wait, why do we want to be generating missing patterns?

File: grammars/silver/compiler/extension/patternmatching/Case.sv
Comment (krame505) [inline grammars/silver/compiler/extension/patternmatching/Case.sv:714]: Could be `cond.isJust`.

File: grammars/silver/compiler/extension/patternmatching/PatternTypes.sv
Comment (krame505) [inline grammars/silver/compiler/extension/patternmatching/PatternTypes.sv:95]: Why is this a "local function", since this is only used in one place?  Took me a minute to understand what was going on here.

File: grammars/silver/compiler/extension/patternmatching/Case.sv
Comment (RandomActsOfGrammar) [reply grammars/silver/compiler/extension/patternmatching/Case.sv:257]: It looked much nicer when compilation was done by iterative forwarding in the production, because the `[Message]` part was added into the `errors` attribute directly.  However, since we can only find the error for overlapping patterns by compiling down to no patterns, and compilation by iterative forwarding breaks completeness analysis, I think this is what we are going to have to use.

The other option, which we decided against when I moved compilation, is to have a compilation production.  This message would then just be part of the `errors` on that.

File: grammars/silver/compiler/extension/patternmatching/Case.sv
Comment (RandomActsOfGrammar) [reply grammars/silver/compiler/extension/patternmatching/Case.sv:426]: It generates an example of a value which is not covered by the given patterns, as well as checking that all the patterns are the same kind of primitive before generating the missing pattern.

If we have patterns `0`, `1`, and `5`, it will generate an example missing pattern of `2`, which will be shown in the warning message.  This is the same behavior as OCaml has.

File: grammars/silver/compiler/extension/patternmatching/PatternTypes.sv
Comment (RandomActsOfGrammar) [reply grammars/silver/compiler/extension/patternmatching/PatternTypes.sv:95]: It is a recursive function to get the absolute head name, so it needs to be a function rather than just matching.

Comment (krame505) [reply grammars/silver/compiler/extension/patternmatching/PatternTypes.sv:95]: Couldn't you just do `case prod.lookupValue.typeScheme.typerep.outputType.baseType of ...`, though? (`.baseType` extracts the applied constructor type at the base of the `appType` chain.)

Comment (RandomActsOfGrammar) [reply grammars/silver/compiler/extension/patternmatching/PatternTypes.sv]: I wasn't aware of the `baseType` attribute; I've changed this to use that now.

File: grammars/silver/compiler/extension/patternmatching/Case.sv
Comment (krame505) [reply grammars/silver/compiler/extension/patternmatching/Case.sv:428]: Ah cool, makes sense.

File: grammars/silver/compiler/extension/patternmatching/PatternTypes.sv
Comment (krame505) [inline grammars/silver/compiler/extension/patternmatching/PatternTypes.sv:94]: Oh, this could probably also just be `.typeName` instead of pattern matching here.

**Review by krame505 — APPROVED** [review]
Approved - pending the downstream builds passing, of course.

**Comment by krame505** [comment]
https://foundry.remexre.xyz/jenkins/job/melt-umn/job/ableC/job/develop/288/console

Is this a bug?
```
 [grammars/edu.umn.cs.melt.ableC/abstractsyntax/host/Stmt.sv]
Stmt.sv:284:16: warning: This pattern-matching is not exhaustive.  Here is an example of a case that is not matched:  _, silver:core:just(_)
```

**Comment by krame505** [comment]
@RandomActsOfGrammar what is the status of this?  I took a look at fixing the case completeness errors in ableC, but the bug I mentioned above hasn't been fixed yet?

**Comment by krame505** [comment]
No particular need to rush, but I do want to get this merged sooner rather than later as there are currently 8 open Silver pull requests.

**Comment by RandomActsOfGrammar** [comment]
I've been working on other things, including changing implicit monads to use type classes.

I have an idea of what the bug is, but I haven't looked into fixing it yet.

**Comment by RandomActsOfGrammar** [comment]
The bug in checking was due to a problem with grouping the patterns.  It turns out the `groupBy` function from `core` doesn't group everything in the list which is related together, only the first elements, so I wrote my own grouping function.

I'll take a look at the incomplete matches in Silver tomorrow, and see if there are any I can fix.

**Comment by krame505** [comment]
In fixing some of the completeness errors I've noticed that the counterexamples could sometimes be more specific.  For example the following case:
```
        case decE1.isSimple, decE2.isSimple, dcls of
        | true, true, _ -> ...
        | true, false, seqStmt(_, d) -> ...
        | false, true, seqStmt(d, _) -> ...
        | false, false, _ -> ...
        end
```
gives the error `This pattern-matching is not exhaustive.  Here is an example of a case that is not matched:  _, _, edu:umn:cs:melt:ableC:abstractsyntax:host:injectGlobalDeclsStmt(_, _)`
Note that there are some ways that the `_`s could be filled in that would match successfully.  I would expect the counterexample to be something like ` true, true, edu:umn:cs:melt:ableC:abstractsyntax:host:injectGlobalDeclsStmt(_, _)`

**Review by krame505 — CHANGES_REQUESTED** [review]
I fixed all the downstream incomplete matches, but I also noticed that convenience aspects now give bogus incomplete match warnings - see the results of building the `silver_features` testsuite.  Note that the top-level of these patterns matches should be handled by the MWDA, but we do want to warn about incomplete matches involving nested patterns since these do become cases in the translation.  

Should be fixed before this is merged, I think.

**Comment by hypnoticOxbow** [comment]
Ah, so I looked at the case completeness for convenience aspects and it looks like those were legitimate issues with the case completeness for the generated aspects, caused by not providing a pattern for a case that is possible with the attribute. I have a fix for the test code available, and can push that here if you'd like.

**Comment by krame505** [comment]
Yeah, go ahead and commit that - thanks.

**Comment by krame505** [comment]
Huh, so
```
aspect bagList2 on top::BazExpr using <- of
| bazInit2(h::t,value) -> [h, toString(value)]
| coolName -> [coolName.hiddenAttr]
end;
```
is an incomplete match despite the default case, because the first thing has a nested pattern, and once it matches `bazInit2` we are in the equation for a particular production, and don't have an option to fall back to a different one.  That's a little weird, but I guess makes sense for what we are trying to express here.  

Do you have a test for a convenience aspect that doesn't cover all productions and doesn't have a default case gives a MWDA error rather than a match completeness error?  Or maybe rather that you can split up the equations for an attribute on a nonterminal across multiple `aspect` declarations without incomplete match warnings?

**Comment by hypnoticOxbow** [comment]
No, I don't have a test for that, but could add one. The only concern is that I'm not certain that the mwda flag is passed for the silver_features tests, so the warning might not appear.

**Comment by krame505** [comment]
Yeah, I meant mainly just testing that an incomplete case at the top-level doesn't produce an incomplete match error.

**Comment by hypnoticOxbow** [comment]
To clarify, you're referring to cases like this right, where if we say there are 4 productions BazInit{1,2,3,4}, and we provide a new attribute and a convenience aspect like the following.
```
aspect bagList on top::BazExpr using <- of
| bazInit2(h::t,value) -> [h, toString(value)]
| bazInit3(_,val) -> [top.hiddenAttr]
| bazInit4() -> explode(top.hiddenAttr,"\t")
end;
```
where in this case we're missing one equation (bazInit1), and your question is whether this will produce an incomplete cases error? Atm the answer appears to be "no", and I can add this as a test case if you'd like.

**Comment by krame505** [comment]
Yes, that is what I was wondering - e.g. if one decided to supply an equation for `bazInit1` with a regular aspect production and the rest in a convenience aspect.  If you could add a test for that it would be great.

**Comment by hypnoticOxbow** [comment]
Sure, I can do that. Should I commit in this branch or make another PR?

**Comment by krame505** [comment]
This branch is fine.

**Comment by ericvanwyk** [comment]
@RandomActsOfGrammar - don't we handle this kind of situation in regular patterns by duplicating, in this case, the default case?

> Huh, so
> 
> ```
> aspect bagList2 on top::BazExpr using <- of
> | bazInit2(h::t,value) -> [h, toString(value)]
> | coolName -> [coolName.hiddenAttr]
> end;
> ```
> 
> is an incomplete match despite the default case, because the first thing has a nested pattern, and once it matches `bazInit2` we are in the equation for a particular production, and don't have an option to fall back to a different one. That's a little weird, but I guess makes sense for what we are trying to express here.
> 
> Do you have a test for a convenience aspect that doesn't cover all productions and doesn't have a default case gives a MWDA error rather than a match completeness error? Or maybe rather that you can split up the equations for an attribute on a nonterminal across multiple `aspect` declarations without incomplete match warnings?

**Comment by RandomActsOfGrammar** [comment]
In a normal `case` expression, when the first pattern didn't match, the default case would be taken.

The issue here is whether that behavior is the right behavior for the aspect productions.  Do we expect default cases only to be used for equations in productions which weren't otherwise specified, or to be used as the default if the patterns for a production weren't complete?

**Comment by krame505** [comment]
I suppose it would make sense to plug in the top-level default case as the default in each aspect production case if none is provided.  

But that is a separate issue from this pull request; I think this is ready to merge now, pending if @RandomActsOfGrammar was going to try fixing the reported counterexamples to be more specific.

**Comment by krame505** [comment]
I'd like to get this merged quickly as there are several other branches waiting for this one to be merged - so I think I'm going to open a new issue for the counterexample issues and merge this now, unless @RandomActsOfGrammar has any objections.

**Review by krame505 — APPROVED** [review]

**Comment by RandomActsOfGrammar** [comment]
I do object.  I found a bug while looking at the counterexample issue.  I'm currently working on it, and hope to have it done tonight.

**Comment by krame505** [comment]
Ok, that's fine with me.

**Comment by RandomActsOfGrammar** [comment]
I ended up having to rearrange the checking quite a bit to fix it.  Previously, it wasn't handling completeness checking with subpatterns correctly and missed some incomplete matches.  There are some incomplete matches in `grammars/silver/util/deque/` to be fixed which were previously missed because of this bug.  There could end up being some in AbleC as well.

This should improve the error messages for missing patterns some as well.  However, if a set of patterns is incomplete overall but an early pattern is completed by a variable, it can still show a variable instead of a missing value.  For example, the patterns
```
| 1, 2, _ -> "first" 
| _, _, 8 -> "second"
```
give a missing pattern `1, _, 0`.  If we fill in `2`, it is covered, but it isn't for any other value.  In the interest of getting this merged, I'll leave looking at that separate and handle it as #479.

File: grammars/silver/compiler/extension/patternmatching/Case.sv
Comment (krame505) [inline grammars/silver/compiler/extension/patternmatching/Case.sv:335]: For future reference, this sort of nested indentation with `if`s is something to avoid as it makes refactoring much more annoying.  We would typically write this as 
```
if numPatts == 0 || !allSameLen
then nothing()
else if isBoolPatts
then boolComp
else if isListPatts
then listComp
else ...
```
indenting the `then` clauses to be the same as the above `else` clauses.

**Comment by krame505** [comment]
Looks like there was just one incomplete match that showed up after you fixed the bug, in the ableC-prolog extension.  I fixed it and will merge this as soon as the build passes on Jenkins.

**Pull request merged**
